// @effect/platform@0.58.26/dist/cjs/internal/workerRunner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeSerialized = exports.make = exports.layerSerialized = exports.layer = exports.PlatformRunnerTypeId = exports.PlatformRunner = void 0;
var Schema = _interopRequireWildcard(require("@effect/schema/Schema"));
var Serializable = _interopRequireWildcard(require("@effect/schema/Serializable"));
var Cause = _interopRequireWildcard(require("effect/Cause"));
var Chunk = _interopRequireWildcard(require("effect/Chunk"));
var Context = _interopRequireWildcard(require("effect/Context"));
var Effect = _interopRequireWildcard(require("effect/Effect"));
var Either = _interopRequireWildcard(require("effect/Either"));
var ExecutionStrategy = _interopRequireWildcard(require("effect/ExecutionStrategy"));
var Exit = _interopRequireWildcard(require("effect/Exit"));
var Fiber = _interopRequireWildcard(require("effect/Fiber"));
var _Function = require("effect/Function");
var Layer = _interopRequireWildcard(require("effect/Layer"));
var Option = _interopRequireWildcard(require("effect/Option"));
var Queue = _interopRequireWildcard(require("effect/Queue"));
var Scope = _interopRequireWildcard(require("effect/Scope"));
var Stream = _interopRequireWildcard(require("effect/Stream"));
var Transferable = _interopRequireWildcard(require("../Transferable.js"));
var _WorkerError = require("../WorkerError.js");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/** @internal */
const PlatformRunnerTypeId = exports.PlatformRunnerTypeId = /*#__PURE__*/Symbol.for("@effect/platform/Runner/PlatformRunner");
/** @internal */
const PlatformRunner = exports.PlatformRunner = /*#__PURE__*/Context.GenericTag("@effect/platform/Runner/PlatformRunner");
/** @internal */
const make = (process, options) => Effect.gen(function* (_) {
  const scope = yield* _(Scope.fork(yield* _(Effect.scope), ExecutionStrategy.parallel));
  const fiber = Option.getOrThrow(Fiber.getCurrentFiber());
  const shutdown = Effect.zipRight(Scope.close(scope, Exit.void), Fiber.interruptFork(fiber));
  const platform = yield* _(PlatformRunner);
  const backing = yield* _(platform.start(shutdown), Scope.extend(scope));
  const fiberMap = new Map();
  yield* _(Queue.take(backing.queue), options?.decode ? Effect.flatMap(msg => {
    const req = msg[1];
    if (req[1] === 1) {
      return Effect.succeed(msg);
    }
    return Effect.map(options.decode(req[2]), data => [msg[0], [req[0], req[1], data, req[3]]]);
  }) : _Function.identity, Effect.tap(([portId, req]) => {
    const id = req[0];
    if (req[1] === 1) {
      const fiber = fiberMap.get(id);
      if (!fiber) return Effect.void;
      return Fiber.interrupt(fiber);
    }
    const collector = Transferable.unsafeMakeCollector();
    return (0, _Function.pipe)(Effect.sync(() => process(req[2])), Effect.flatMap(stream => {
      let effect = Effect.isEffect(stream) ? Effect.flatMap(stream, data => {
        const transfers = options?.transfers ? options.transfers(data) : [];
        return (0, _Function.pipe)(options?.encodeOutput ? Effect.provideService(options.encodeOutput(req[2], data), Transferable.Collector, collector) : Effect.succeed(data), Effect.flatMap(payload => backing.send(portId, [id, 0, [payload]], [...transfers, ...collector.unsafeRead()])));
      }) : (0, _Function.pipe)(stream, Stream.chunks, Stream.tap(data => {
        if (options?.encodeOutput === undefined) {
          const payload = Chunk.toReadonlyArray(data);
          const transfers = options?.transfers ? payload.flatMap(options.transfers) : undefined;
          return backing.send(portId, [id, 0, payload], transfers);
        }
        const transfers = [];
        collector.unsafeClear();
        return (0, _Function.pipe)(Effect.forEach(data, data => {
          if (options?.transfers) {
            for (const option of options.transfers(data)) {
              transfers.push(option);
            }
          }
          return Effect.orDie(options.encodeOutput(req[2], data));
        }), Effect.provideService(Transferable.Collector, collector), Effect.flatMap(payload => {
          collector.unsafeRead().forEach(transfer => transfers.push(transfer));
          return backing.send(portId, [id, 0, payload], transfers);
        }));
      }), Stream.runDrain, Effect.andThen(backing.send(portId, [id, 1])));
      if (req[3]) {
        const [traceId, spanId, sampled] = req[3];
        effect = Effect.withParentSpan(effect, {
          _tag: "ExternalSpan",
          traceId,
          spanId,
          sampled,
          context: Context.empty()
        });
      }
      return effect;
    }), Effect.catchIf(_WorkerError.isWorkerError, error => backing.send(portId, [id, 3, _WorkerError.WorkerError.encodeCause(Cause.fail(error))])), Effect.catchAllCause(cause => Either.match(Cause.failureOrCause(cause), {
      onLeft: error => {
        const transfers = options?.transfers ? options.transfers(error) : [];
        collector.unsafeClear();
        return (0, _Function.pipe)(options?.encodeError ? Effect.provideService(options.encodeError(req[2], error), Transferable.Collector, collector) : Effect.succeed(error), Effect.flatMap(payload => backing.send(portId, [id, 2, payload], [...transfers, ...collector.unsafeRead()])), Effect.catchAllCause(cause => backing.send(portId, [id, 3, _WorkerError.WorkerError.encodeCause(cause)])));
      },
      onRight: cause => backing.send(portId, [id, 3, _WorkerError.WorkerError.encodeCause(cause)])
    })), Effect.ensuring(Effect.sync(() => fiberMap.delete(id))), Effect.fork, Effect.tap(fiber => Effect.sync(() => fiberMap.set(id, fiber))));
  }), Effect.forever, Effect.forkIn(scope));
});
/** @internal */
exports.make = make;
const layer = (process, options) => Layer.scopedDiscard(make(process, options));
/** @internal */
exports.layer = layer;
const makeSerialized = (schema, handlers) => Effect.gen(function* (_) {
  const scope = yield* _(Effect.scope);
  let context = Context.empty();
  const parseRequest = Schema.decodeUnknown(schema);
  return yield* _(make(request => {
    const result = handlers[request._tag](request);
    if (Layer.isLayer(result)) {
      return Effect.flatMap(Layer.buildWithScope(result, scope), _ => Effect.sync(() => {
        context = Context.merge(context, _);
      }));
    } else if (Effect.isEffect(result)) {
      return Effect.provide(result, context);
    }
    return Stream.provideContext(result, context);
  }, {
    decode(message) {
      return Effect.mapError(parseRequest(message), error => new _WorkerError.WorkerError({
        reason: "decode",
        error
      }));
    },
    encodeError(request, message) {
      return Effect.mapError(Serializable.serializeFailure(request, message), error => new _WorkerError.WorkerError({
        reason: "encode",
        error
      }));
    },
    encodeOutput(request, message) {
      return Effect.catchAllCause(Serializable.serializeSuccess(request, message), error => new _WorkerError.WorkerError({
        reason: "encode",
        error
      }));
    }
  }));
});
/** @internal */
exports.makeSerialized = makeSerialized;
const layerSerialized = (schema, handlers) => Layer.scopedDiscard(makeSerialized(schema, handlers));
exports.layerSerialized = layerSerialized;
