// @effect/platform@0.58.26/dist/cjs/WorkerError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isWorkerError = exports.WorkerErrorTypeId = exports.WorkerError = void 0;
var Schema = _interopRequireWildcard(require("@effect/schema/Schema"));
var Inspectable = _interopRequireWildcard(require("effect/Inspectable"));
var Predicate = _interopRequireWildcard(require("effect/Predicate"));
var internal = _interopRequireWildcard(require("./internal/workerError.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category type ids
 */
const WorkerErrorTypeId = exports.WorkerErrorTypeId = internal.WorkerErrorTypeId;
/**
 * @since 1.0.0
 * @category predicates
 */
const isWorkerError = u => Predicate.hasProperty(u, WorkerErrorTypeId);
/**
 * @since 1.0.0
 * @category errors
 */
exports.isWorkerError = isWorkerError;
class WorkerError extends Schema.TaggedError()("WorkerError", {
  reason: Schema.Literal("spawn", "decode", "send", "unknown", "encode"),
  error: Schema.CauseDefectUnknown
}) {
  /**
   * @since 1.0.0
   */
  [WorkerErrorTypeId] = WorkerErrorTypeId;
  /**
   * @since 1.0.0
   */
  static Cause = Schema.Cause({
    error: this
  });
  /**
   * @since 1.0.0
   */
  static encodeCause = Schema.encodeSync(this.Cause);
  /**
   * @since 1.0.0
   */
  static decodeCause = Schema.decodeSync(this.Cause);
  /**
   * @since 1.0.0
   */
  get message() {
    const message = Predicate.hasProperty(this.error, "message") ? this.error.message : Inspectable.toStringUnknown(this.error, undefined);
    return `${this.reason}: ${message}`;
  }
}
exports.WorkerError = WorkerError;
